script/tool/lib/src/update_release_info_command.dart

```diff
@@ -0,0 +1,303 @@
+        _changedFiles = await gitVersionFinder.getChangedFiles();
+        // Anothing other than a fixed change is null.
+        _versionChange = null;
+        break;
+      case _versionNext:
+        _versionChange = null;
+        break;
+      default:
+        throw UnimplementedError('Unimplemented version change type');
+    }
+  }
+
+  @override
+  Future<PackageResult> runForPackage(RepositoryPackage package) async {
+    String nextVersionString;
+
+    _VersionIncrementType? versionChange = _versionChange;
+    // If the change type is `minimal`, determine whether a version change is
+    // needed.
+    if (versionChange == null &&
+        getStringArg(_versionTypeFlag) == _versionMinimal) {
+      final Directory gitRoot =
+          packagesDir.fileSystem.directory((await gitDir).path);
+      final String relativePackagePath =
+          getRelativePosixPath(package.directory, from: gitRoot);
+      final PackageChangeState state = checkPackageChangeState(package,
+          changedPaths: _changedFiles,
+          relativePackagePath: relativePackagePath);
+      if (state.needsVersionChange) {
+        versionChange = _VersionIncrementType.bugfix;
+      }
+    }
+
+    if (versionChange != null) {
+      final Version? updatedVersion =
+          _updatePubspecVersion(package, versionChange);
+      if (updatedVersion == null) {
+        return PackageResult.fail(
+            <String>['Could not determine current version.']);
+      }
+      nextVersionString = updatedVersion.toString();
+      print('${indentation}Incremented version to $nextVersionString.');
+    } else {
+      nextVersionString = 'NEXT';
+    }
+
+    final _ChangelogUpdateOutcome updateOutcome =
+        _updateChangelog(package, nextVersionString);
+    switch (updateOutcome) {
+      case _ChangelogUpdateOutcome.addedSection:
+        print('${indentation}Added a $nextVersionString section.');
+        break;
+      case _ChangelogUpdateOutcome.updatedSection:
+        print('${indentation}Updated NEXT section.');
+        break;
+      case _ChangelogUpdateOutcome.failed:
+        return PackageResult.fail(<String>['Could not update CHANGELOG.md.']);
+    }
+
+    return PackageResult.success();
+  }
+
+  _ChangelogUpdateOutcome _updateChangelog(
+      RepositoryPackage package, String version) {
+    if (!package.changelogFile.existsSync()) {
+      printError('${indentation}Missing CHANGELOG.md.');
+      return _ChangelogUpdateOutcome.failed;
+    }
+
+    final String newHeader = '## $version';
+    final RegExp listItemPattern = RegExp(r'^(\s*[-*])');
+
+    final StringBuffer newChangelog = StringBuffer();
+    _ChangelogUpdateState state = _ChangelogUpdateState.findingFirstSection;
+    bool updatedExistingSection = false;
+
+    for (final String line in package.changelogFile.readAsLinesSync()) {
+      switch (state) {
+        case _ChangelogUpdateState.findingFirstSection:
```

Comment:
Can you elaborate on what you had in mind here? This is a very simple state machine, and it currently relies on some shared state (new version, new changelog) and logic (list building given the found marker) from its enclosing class.

I'm not seeing offhand how to apply the state pattern without ending up with at least four classes that are constantly passing state amongst themselves to preserve that shared information, and/or defining a delegation interface just to get that on demand (and then still having to pass around a delegate object). Doing that would be something like 1.5-2x the code for the state machine, with pretty much all of the additional code being boilerplate that would obscure the logic. And I'm not clear on what the benefit we'd get to maintainability here given the small size and fixed scope (there's almost no chance we will be adding new changelog complexity since it's a well established format).

Am I misunderstanding the suggestion?